components/authentication-framework/org.wso2.carbon.identity.application.authentication.framework/src/main/java/org/wso2/carbon/identity/application/authentication/framework/handler/provisioning/ProvisionedUserDeleteThread.java

```diff
@@ -0,0 +1,82 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.wso2.carbon.identity.application.authentication.framework.handler.provisioning;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.wso2.carbon.identity.application.authentication.framework.internal.FrameworkServiceComponent;
+import org.wso2.carbon.identity.application.authentication.framework.store.SessionCleanUpService;
+import org.wso2.carbon.identity.application.authentication.framework.util.FrameworkConstants;
+import org.wso2.carbon.identity.application.authentication.framework.util.FrameworkUtils;
+import org.wso2.carbon.user.api.UserRealm;
+import org.wso2.carbon.user.api.UserStoreException;
+import org.wso2.carbon.user.core.common.AbstractUserStoreManager;
+import org.wso2.carbon.user.core.service.RealmService;
+
+/**
+ * Thread to perform JIT provisioned users deletion task based on provisioned IDP delete.
+ */
+public class ProvisionedUserDeleteThread implements Runnable {
+
+    private static final Log log = LogFactory.getLog(SessionCleanUpService.class);
+    private static final Log diagnosticLog = LogFactory.getLog("diagnostics");
+
+    private final String resourceId;
+    private final String tenantDomain;
+
+    public ProvisionedUserDeleteThread(String resourceId, String tenantDomain) {
+
+        this.resourceId = resourceId;
+        this.tenantDomain = tenantDomain;
+    }
+
+    @Override
+    public void run() {
+
+        if (log.isDebugEnabled()) {
+            log.debug("Start running the JIT provisioned user delete task.");
+        }
+        diagnosticLog.info("Start running the JIT provisioned user delete task.");
+        provisionedUserDelete();
+        if (log.isDebugEnabled()) {
+            log.debug("Stop running the JIT provisioned user delete task.");
+        }
+        diagnosticLog.info("Stop running the JIT provisioned user delete task.");
+    }
+
+    private void provisionedUserDelete() {
+
+        try {
+            FrameworkUtils.startTenantFlow(tenantDomain);
+            RealmService realmService = FrameworkServiceComponent.getRealmService();
+            int tenantId = realmService.getTenantManager().getTenantId(tenantDomain);
+            UserRealm userRealm = realmService.getTenantUserRealm(tenantId);
+            String[] userList =
+                    ((AbstractUserStoreManager)userRealm.getUserStoreManager()).
+                            getUserList(FrameworkConstants.PROVISIONED_SOURCE_CLAIM, resourceId, null);
+            for (String username : userList) {
```

Comment:
Shall we stop before the `for loop` if we returned a null or empty list

Reply:
Fixed https://github.com/wso2/carbon-identity-framework/pull/3597/commits/6e1589981f43782e7824b22d065010c4ef34b01d